### PR TITLE
Add exit pages for Trainees and Teachers

### PR DESCRIPTION
**Pull request by padv2010**
## Trello Card Link

https://trello.com/c/fsy6rp4P/72-teacher-route-exit-page-copy-application-criteria-is-fulfilled
https://trello.com/c/DfypaSU8/71-trainee-exit-page-copy-application-criteria-is-fulfilled
https://trello.com/c/acwINfog/32-m-generate-urn-for-applications-on-submission

## Description

Add the exit page for trainee and teachers. Both routes use the same copy for the content. I have used the [GOV.UK guidelines for Applications](https://design-system.service.gov.uk/patterns/confirmation-pages/).

## Technical notes

1. Introduces a fix in 89e990e44502b52d211b8d2f3e41a3db93e81398 for #63 because I didn't wrap the value with the `Urn` class to get the most of encapsulation.
2. Introduces a new `request` test for the Personal Details question. This is now needed because I made some changes to how data is persisted.
3. Introduces the `Application` model following up #73. Extracting this domain model enables an easy location for the `urn` property that is displayed on the exit page.
4. In 1f8ed820699118a8b00c566a8f31908320bd894c introduces a controller for rendering the exit page. This enables rendering the `urn` for the application in a consistent way with the other questions.
5. It moves persisting attributes out of the controllers for: personal details and employment details. I have moved them into a `save!` method at the domain level. As suggested in https://github.com/DFE-Digital/international-teacher-relocation-payment/pull/70#discussion_r1234018276, it is a good thing to move into the `forms` domain folder, but I would rather doing it with all the classes and not with two in a single PR for consistency. I will do at a later stage as this PR is too big to include those changes.
6. It leverages ActiveRecord persistence and [dries-up controllers](https://github.com/DFE-Digital/international-teacher-relocation-payment/compare/exit-page-application-model?expand=1#diff-171fa0c697ee861041d20cf7716741b0f3e4ed647893a0379e5ce9031402ea69L38-L80) using the session only for the `applicant_id`. 

## Screenshots

image

### CODE REVIEW

**Comment by padv2010** [comment]
[Teacher route: Exit page copy (application criteria is fulfilled)](https://trello.com/c/fsy6rp4P/72-teacher-route-exit-page-copy-application-criteria-is-fulfilled)

**Comment by padv2010** [comment]
[Trainee: Exit page copy (application criteria is fulfilled)](https://trello.com/c/DfypaSU8/71-trainee-exit-page-copy-application-criteria-is-fulfilled)

**Comment by padv2010** [comment]
[[M] Generate URN for applications on submission](https://trello.com/c/acwINfog/32-m-generate-urn-for-applications-on-submission)

File: spec/factories/applications.rb [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+#  applicant_id     :bigint           not null
+#
+# Foreign Keys
+#
+#  fk_rails_...  (applicant_id => applicants.id)
+#
+FactoryBot.define do
+  factory :application do
+    application_date { Faker::Date.in_date_period }
+    urn { "IRP #{Faker::Number.number(digits: 6)} LT" }
```
Comment (duncanjbrown) [inline spec/factories/applications.rb:19]: Why not use `Urn.new.value`? & What's the LT?

File: spec/models/urn_spec.rb [hunk 2]
```diff
@@ -13,22 +13,22 @@
...
     charset = %w[A B C D E F H J K L M N P R S T U V 0 1 2 3 4 5 6 7 8 9]
 
-    expect(urn.value[3..].chars).to all(be_in(charset))
+    expect(urn.value[4..].chars).to all(be_in(charset))
   end
 
   describe ".dump" do
     it "returns the given value" do
-      expect(described_class.dump("IRPG2345")).to eq("IRPG2345")
+      expect(described_class.dump(described_class.new(value: "IRPG2345"))).to eq("IRPG2345")
```
Comment (duncanjbrown) [inline spec/models/urn_spec.rb:31]: shall we consistently space between `IRP` and the digits, just for 📐 ?

File: app/controllers/applicants/submissions_controller.rb [hunk 3]
```diff
@@ -0,0 +1,7 @@
+# frozen_string_literal: true
+
+module Applicants
+  class SubmissionsController < ApplicationController
+    def show; end
```
Comment (duncanjbrown) [inline app/controllers/applicants/submissions_controller.rb]: 👌

File: app/views/applicants/submissions/show.html.erb [hunk 4]
```diff
@@ -10,7 +10,7 @@
 <div class="govuk-grid-row">
   <div class="govuk-grid-column-two-thirds-from-desktop">
     <p class="govuk-body">
-      Your details have been submitted and will now be checked by the
+     Your details have been submitted and will now be checked by the
```
Comment (duncanjbrown) [inline app/views/applicants/submissions/show.html.erb]: Rogue backspace?

File: app/views/applicants/submissions/show.html.erb [hunk 5]
```diff
@@ -1,28 +1,40 @@
...
+        </p>
+
+        <p class="govuk-body">
+          We'll let you know the outcome of your application as soon as we've checked your eligibility. If you are eligible,
+          you will be paid around the end of your first term of training or employment.
+        </p>
+
+        <p class="govuk-body">
+          For help, you can email us at
+          <%= govuk_link_to("teach.inengland@education.gov.uk", "mailto:teach.inengland@education.gov.uk") %>.
```
Comment (duncanjbrown) [inline app/views/applicants/submissions/show.html.erb:30]: btw we have these cops in other apps, which are good to get in early! https://github.com/DFE-Digital/apply-for-teacher-training/tree/main/lib/rubocop/cop/govuk

File: app/controllers/applicants/submissions_controller.rb [hunk 6]
```diff
@@ -3,7 +3,9 @@
 module Applicants
   class SubmissionsController < ApplicationController
     def show
-      @applicant = Applicant.find(session[:applicant_id])
+      applicant = Applicant.find(session[:applicant_id])
```
Comment (duncanjbrown) [inline app/controllers/applicants/submissions_controller.rb:6]: Could go to `current_applicant` at this point, then `@application = current_applicant.application`?

File: app/views/applicants/submissions/show.html.erb [hunk 7]
```diff
@@ -7,7 +7,7 @@
             Application complete
           </h1>
           <div class="govuk-panel__body">
-            Your reference number<br><strong><%= @applicant.application.urn.value %></strong>
+            Your reference number<br><strong><%= @application.urn.value %></strong>
```
Comment (duncanjbrown) [inline app/views/applicants/submissions/show.html.erb:10]: 👌

File: app/controllers/applicants/employment_details_controller.rb [hunk 8]
```diff
@@ -15,7 +15,7 @@ def create
         @employment_detail.applicant = @applicant
         @employment_detail.save!
 
-        @applicant.create_application!(application_date: Date.current)
+        Application.register_for_applicant!(@applicant)
```
Comment (duncanjbrown) [inline app/controllers/applicants/employment_details_controller.rb:18]: What does 'register' mean in this context? Is it different from 'initialise'?

**Review by duncanjbrown — APPROVED** [review]
🙌 Looks great — (minor) comments inline!

File: app/views/applicants/submissions/show.html.erb (hunk 5 above)
Comment (padv2010) [reply app/views/applicants/submissions/show.html.erb:30]: I am [tracking this work on Trello](https://trello.com/c/youTNukD/90-add-govuk-cops) and I aim to address it next.

File: app/controllers/applicants/submissions_controller.rb (hunk 6 above)
Comment (padv2010) [reply app/controllers/applicants/submissions_controller.rb]: @duncanjbrown addressed in 4201464. Code looks much better, thanks 👍

File: app/controllers/applicants/employment_details_controller.rb (hunk 8 above)
Comment (padv2010) [reply app/controllers/applicants/employment_details_controller.rb]: @duncanjbrown thanks for feedback. `register` is probably the wrong name. I have tried with `initialise_for_applicant` and reads better on 593abde.

**Pull request merged**